Dry up search routes

wwilsman opened the pull request:
While the size diff remains about the same, this PR pulls all search logic into a single search component and creates a reducer and epic specific to the search type (`vendors`, `packages`, `titles`).

I've also introduced some things that might make changing query params later easier, such as keeping the current query in the state for that search type. This allows us to restore a previous search when navigating around the app.

Since ejecting from `stripes-connect` we can now handle errors better also, although as of right now there is no UI around handling search-related errors.

### Screenshots
![keep track of searches](https://user-images.githubusercontent.com/5005153/30131347-1607e42e-9312-11e7-9370-ff24d72cfaf9.gif)

wwilsman commented on src/redux/search.js:
```diff
@@ -0,0 +1,92 @@
+
+        return Observable.from(request.then((res) => res.json())).pluck(recordsKey)
+          .map((payload) => ({ type: EHOLDINGS_SEARCH_RESOLVE, searchType, payload }))
+          .catch((error) => Observable.of({ type: EHOLDINGS_SEARCH_ERRORED, searchType, error }));
```
Just following what `redux-observable` had in their examples. I actually had it the other way until I saw [an example](https://redux-observable.js.org/docs/recipes/ErrorHandling.html)

wwilsman commented on src/redux/search.js:
```diff
@@ -0,0 +1,92 @@
+  };
+
+  return (state = initialState, action) => {
+    if (searchType === action.searchType) {
```
The only thing about using `handleActions` here is that I'd have to create a reference to the reducer it creates and call it as a function inside of the if statement. If there was more logic going on in each action handler besides a single return, I'd agree with `handleActions`.

wwilsman commented on src/index.js:
```diff
@@ -37,9 +34,9 @@ export default class EHoldings extends Component {
-        <Route path={`${rootPath}/vendors`} exact component={this.SearchVendors}/>
-        <Route path={`${rootPath}/packages`} exact component={this.SearchPackages}/>
-        <Route path={`${rootPath}/titles`} exact component={this.SearchTitles}/>
+        <Route path={`${rootPath}/vendors`} exact component={SearchVendors}/>
```
I agree. I almost renamed them but thought it wasn't necessary. I'll adjust

wwilsman commented:
@cowboyd regarding routes and presentation:

I could pull out `renderRecords` and instead have it be something like `resultsComponent` which would receive `records`, `isLoaded`, `isErrored`, etc. via properties.

Then `VendorSearch` and others would simply render and pass relevant props to `Search`. We could even skip the `VendorSearch` component (if we wanted) and just render a `Search` component within the route components:

```js
<Route path={`${rootPath}/vendors`} exact render={(props) => (
  <SearchRoute
      searchType="vendors"
      endpoint="eholdings/vendors"
      resultsComponent={VendorsSearchResults}
      {/* other relevant properties */}
      {...props}/>
)}/>
```

Also this way, each `view` component would be responsible for displaying it's own "not found" and "loading" messages.

And of course the render methods inside the actual `Search` component could be pulled out into their own set of components.

sadatay commented on src/redux/search.js:
```diff
@@ -0,0 +1,92 @@
+        const { okapi } = getState();
+
+        const request = fetch(getSearchURL(okapi.url, action.query), {
+          headers: { 'X-Okapi-Tenant': okapi.tenant }
```
we do this a lot.  i wonder if there's a way to apply the okapi header (soon to be headers) wholesale...

wwilsman replied:
We could do something like define a helper such as `fetchFromOkapi(okapiObject, resourceEndpoint, fetchOptions)` and construct the headers there

wwilsman commented:
Updated the reducer to be specific to searches instead of search types and stores each type in a separate key in the state. With this comes a single search component that handles the different search types, with the results components being separate routes connected to the store themselves.

I also decided to go ahead and create a root reducer and epic using `combineReducers` and `combineEpics` to initialize our reducers and epics all at once when the app is mounted. This way other routes can access different pieces of our entire app state by `connect`ing to the `eholdings` slice of the state.

```js
connect(({ eholdings } => ({ ...eholdings.search }))(SearchRoute);
```

This means that the recent work around selecting and deselecting titles will have to be refactored to utilized our app's root reducer / epic.

While eliminating CSS from the routes, I noticed that a lot of list styles were being repeated in literally every other component. So I also made a single `List` component and made use of it everywhere it's needed. In turn, this made it so a lot of components only had a single file in their respective directories, so I've moved those into the `components` root
